contracts/SimpleGAMultiSig.aes

```diff
@@ -0,0 +1,153 @@
+// Currently only one proposed transaction can be multisigned at a time
+contract SimpleGAMultiSig =
+
+  datatype event
+    = TxProposed(hash, address, int)
+    | TxConfirmed(hash, address)
+    | TxConsensusReached(hash)
+    | TxAuthorized(hash)
+    | TxRevoked(address)
+
+  record transaction = 
+    { ga_tx_hash : hash,
+      expires_at : int,
+      proposed_by : address,
+      confirmed_by : list(address) }
+
+  record state = 
+  	{ multisig_users : list(address),
+  	  confirmations_required : int,    
+      current_tx : option(transaction),
+      nonce : int,
+      owner: address }
+
+  stateful entrypoint init(p_confirmations_required : int, p_multisig_users : list(address)) = 
+    { multisig_users = Call.caller :: p_multisig_users, 
+  	  confirmations_required = p_confirmations_required,
+  	  current_tx = None,
+  	  nonce = 1,
+      owner = Call.caller }
+
+  // GA authorize function
+  stateful entrypoint authorize(sig: signature, n : int) : bool =  
+    check_nonce(n)
+    switch(is_expired())
+      true => 
+        abort("ERROR_TX_EXPIRED")
+      false =>
+        require(has_consensus(), "ERROR_NO_CONSENSUS")
+        switch(Auth.tx_hash)
+          None => abort("ERROR_NO_AUTH_CONTEXT")
+          Some(ga_tx_hash) =>
+            require(ga_tx_hash == Option.force(state.current_tx).ga_tx_hash, "ERROR_UNEQUAL_HASHES")
+            require(is_authorized_and_valid(Call.caller, ga_tx_hash, sig), "ERROR_NOT_AUTHORIZED")
+            Chain.event(TxAuthorized(ga_tx_hash))        
+            true
+    
+  // propose a new tx valid for the given ttl
+  stateful entrypoint propose(ga_tx_hash : hash, ttl : Chain.ttl, sig : signature, n : int) =
+    check_nonce(n)
+    require(is_authorized_and_valid(Call.caller, ga_tx_hash, sig), "ERROR_NOT_AUTHORIZED")
+    switch(state.current_tx)
+      None => set_current_tx(ga_tx_hash, ttl)
+      Some(tx) => 
+        switch(is_expired())
+          false => abort("ERROR_A_TX_IS_ALREADY_PROPOSED")
+          true => set_current_tx(ga_tx_hash, ttl)
+  
+  // signer confirms tx
+  stateful entrypoint confirm(sig : signature, n : int) = 
+    check_nonce(n)
+    switch(state.current_tx)
+      None => abort("ERROR_NOTHING_TO_CONFIRM")
+      Some(tx) =>
+        require(is_authorized_and_valid(Call.caller, tx.ga_tx_hash, sig), "ERROR_NOT_AUTHORIZED")
+        require(!List.contains(Call.caller, tx.confirmed_by),"ERROR_ALREADY_CONFIRMED")
+        require(!is_expired(), "ERROR_TX_ALREADY_EXPIRED")        
+        let updated_tx = tx{confirmed_by=Call.caller :: tx.confirmed_by}
+        put(state{current_tx=Some(updated_tx)})
+        Chain.event(TxConfirmed(tx.ga_tx_hash, Call.caller))
+        if(has_consensus())
+          Chain.event(TxConsensusReached(tx.ga_tx_hash))
+  
+  // revoke the current tx and clean state
+  stateful entrypoint revoke(sig : signature, n: int) =
+    check_nonce(n)
+    switch(state.current_tx)
```

Comment:
no check for is expired?

Reply:
not sure, I think it should be allowed to clean the state by every multisig user independent of expiration. what do you think? if we only allow revocation after expiration we should rename the entrypoint IMO

Reply:
IMO we don't need to check. it's only possible to handle one tx at the same time and every signer is allowed to revoke at any point of time